minicli/__init__.py

```diff
@@ -143,21 +111,90 @@ def cli(*args, **kwargs):
     if hasattr(func, '_cli') and len(args) > 1 and kwargs:
         # Chaining cli(xxx) calls.
         extra = func._cli.extra
-        # We don't know how to update an existing action, so let's reset it
-        # totally with new signature.
-        for idx, action in enumerate(subparsers._choices_actions[:]):
-            if action.dest == func._cli.name:
-                del subparsers._choices_actions[idx]
     if len(args) > 1:
         extra[args[1]] = kwargs
     Cli(func, **extra)
     return func
 
 
-def run(*args):
-    parsed = parser.parse_args(args or None)
+def run(*input, **globals):
+    parser = argparse.ArgumentParser(add_help=False)
+    for name, kwargs in globals.items():
+        if not isinstance(kwargs, dict):
+            kwargs = {'default': kwargs}
+        args, kwargs = make_argument(name, **kwargs)
+        parser.add_argument(*args, **kwargs)
+    # globals must be parsed before actual commands so they can be passed to
+    # before hook
+    parsed, extras = parser.parse_known_args(input or None)
+    globals = {k: getattr(parsed, k, None) for k in globals.keys()
+               if hasattr(parsed, k)}
+    # No command is known when calling parse_known_args, prevent argparse to
+    # display the help and exit.
+    parser.add_argument('-h', '--help', action='store_true',
+                        help='Show this help message and exit')
+    subparsers = parser.add_subparsers(title='Available commands', metavar='')
+    for cmd in REGISTRY:
+        cmd.init_parser(subparsers)
+    parsed = parser.parse_args(args=extras)
     if hasattr(parsed, 'func'):
-        parsed.func(parsed)
+        after_before(BEFORE, **globals)
+        parsed.func(parsed, **globals)
+        after_before(AFTER, **globals)
     else:
         # No argument given, just display help.
         parser.print_help()
+        parser.exit()  # Mimic original behaviour.
+
+
+def before(func):
+    BEFORE.append(func)
+    return func
+
+
+def after(func):
+    AFTER.append(func)
+    return func
+
+
+def after_before(funcs, **globals):
+    for func in funcs:
+        spec = inspect.signature(func)
+        async = inspect.iscoroutinefunction(func)
```

Comment:
💣 reserved keyword?

Reply:
Uh! Fun that it worked. Fixed anyway :)

Reply:
I was surprised too, especially with `if async` ;)